From 673a99520a1c234e4bc7c91801a1b20da6600a43 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 8 Jul 2024 09:59:08 +0200
Subject: [PATCH 1194/1695] plat-stm32mp1: pageable PM sequences

Make stm32mp1 PM system-off and system-suspend sequences to execute
in a pageable thread context when CFG_PAGED_PSCI_SYSTEM_OFF and
CFG_PAGED_PSCI_SYSTEM_SUSPEND are enabled.

Change stm32mp_cpu_reset_state() to call straight the resume entry
point. This is needed because that function disables the MMU, hence
the pager so we can't use the stack to return to caller since
the stack is likely pageable.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 57c0b109ac7067ab2fdf04cbfcccf3227265a80f plat-stm32mp1: pageable PM sequences]
Change-Id: I063b583f1195c716f93a5b802e02b201b9ee8081
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/392006
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/conf.mk           |  4 ++
 .../arm/plat-stm32mp1/link_dummies_paged.c    | 21 +++++++++++
 core/arch/arm/plat-stm32mp1/pm/context.h      |  2 +-
 core/arch/arm/plat-stm32mp1/pm/low_power.c    |  3 +-
 core/arch/arm/plat-stm32mp1/pm/pm_helpers.S   | 29 +++++++++++++--
 core/arch/arm/plat-stm32mp1/pm/psci.c         | 37 +++++++++++++++----
 .../arch/arm/plat-stm32mp1/pm/stm32mp1_psci.h | 19 ++++++++++
 7 files changed, 102 insertions(+), 13 deletions(-)
 create mode 100644 core/arch/arm/plat-stm32mp1/pm/stm32mp1_psci.h

diff --git a/core/arch/arm/plat-stm32mp1/conf.mk b/core/arch/arm/plat-stm32mp1/conf.mk
index 5ad6bb6e6..b499efde9 100644
--- a/core/arch/arm/plat-stm32mp1/conf.mk
+++ b/core/arch/arm/plat-stm32mp1/conf.mk
@@ -252,6 +252,10 @@ endif
 endif # CFG_STM32MP15
 
 ifeq ($(CFG_WITH_PAGER),y)
+# Default use pageable PM sequence to relax pressure on pager
+CFG_PAGED_PSCI_SYSTEM_OFF ?= y
+CFG_PAGED_PSCI_SYSTEM_SUSPEND ?= y
+# Default use 32bit MMU since it consumes a bit less of resident memory
 CFG_WITH_LPAE ?= n
 endif
 CFG_WITH_LPAE ?= y
diff --git a/core/arch/arm/plat-stm32mp1/link_dummies_paged.c b/core/arch/arm/plat-stm32mp1/link_dummies_paged.c
index 989d5c2ec..a325848ff 100644
--- a/core/arch/arm/plat-stm32mp1/link_dummies_paged.c
+++ b/core/arch/arm/plat-stm32mp1/link_dummies_paged.c
@@ -1,9 +1,30 @@
 // SPDX-License-Identifier: BSD-2-Clause
 /*
  * Copyright (c) 2022, Linaro Limited
+ * Copyright (c) 2024, STMicroelectronics
  */
 #include <compiler.h>
 #include <drivers/clk.h>
+#include <kernel/panic.h>
+#include <pm/stm32mp1_psci.h>
 #include <stm32_util.h>
 
 const struct clk_ops stm32mp1_clk_ops __rodata_dummy;
+
+#ifdef CFG_PAGED_PSCI_SYSTEM_SUSPEND
+uint32_t __section(".text.dummy.__psci_system_suspend")
+__psci_system_suspend(uintptr_t entry __unused,
+		      uint32_t context_id __unused,
+		      struct sm_nsec_ctx *nsec __unused)
+{
+	return 0;
+}
+#endif
+
+#ifdef CFG_PAGED_PSCI_SYSTEM_OFF
+void __section(".text.dummy.__psci_system_off") __noreturn
+__psci_system_off(void)
+{
+	panic();
+}
+#endif
diff --git a/core/arch/arm/plat-stm32mp1/pm/context.h b/core/arch/arm/plat-stm32mp1/pm/context.h
index 8fc34af94..1ce85b1f6 100644
--- a/core/arch/arm/plat-stm32mp1/pm/context.h
+++ b/core/arch/arm/plat-stm32mp1/pm/context.h
@@ -79,7 +79,7 @@ extern const uint8_t stm32mp_bkpsram_image_end[];
 void stm32mp_bkpsram_resume(void);
 void stm32mp_sysram_resume(void);
 
-void stm32mp_cpu_reset_state(void);
+void stm32mp_cpu_reset_state(void *resume_ep);
 
 TEE_Result stm32mp_pm_save_context(unsigned int soc_mode);
 void stm32mp_pm_restore_context(unsigned int soc_mode);
diff --git a/core/arch/arm/plat-stm32mp1/pm/low_power.c b/core/arch/arm/plat-stm32mp1/pm/low_power.c
index 9f8026e37..c0c6afc49 100644
--- a/core/arch/arm/plat-stm32mp1/pm/low_power.c
+++ b/core/arch/arm/plat-stm32mp1/pm/low_power.c
@@ -496,8 +496,7 @@ void __noreturn stm32_pm_cpu_power_down_wfi(void)
 
 		/* STANDBY not reached: resume from retained SYSRAM */
 		stm32_exit_cstop();
-		stm32mp_cpu_reset_state();
-		reset_ep();
+		stm32mp_cpu_reset_state(reset_ep);
 		panic();
 	}
 
diff --git a/core/arch/arm/plat-stm32mp1/pm/pm_helpers.S b/core/arch/arm/plat-stm32mp1/pm/pm_helpers.S
index 452535700..6c4a0e16e 100644
--- a/core/arch/arm/plat-stm32mp1/pm/pm_helpers.S
+++ b/core/arch/arm/plat-stm32mp1/pm/pm_helpers.S
@@ -11,6 +11,7 @@
 #include <boot_api.h>
 #include <drivers/stm32mp1_rcc.h>
 #include <generated/context_asm_defines.h>
+#include <keep.h>
 #include <kernel/cache_helpers.h>
 #include <kernel/tz_proc_def.h>
 #include <kernel/tz_ssvce_def.h>
@@ -108,6 +109,19 @@
  */
 FUNC stm32mp_bkpsram_resume , :
 UNWIND(	.cantunwind)
+	/*
+	 * Ensure aligned on a page so that pager does not swap memory during
+	 * the TEE RAM encryption process. With this alignment, page may map
+	 * stm32mp_bkpsram_resume() (that includes the full TEE RAM encryption
+	 * sequence) at entry (stm32mp_ccm_teeram) but will not need to swap
+	 * anything until stm32mp_ccm_teeram is done.
+	 */
+	.balign 4096
+
+#ifdef CFG_PAGED_PSCI_SYSTEM_SUSPEND
+	/* Restore supervisor mode where pager is supported */
+	cps	#CPSR_MODE_SVC
+#endif
 
 	PRINT_CHAR r0, r1, '0'
 
@@ -165,6 +179,13 @@ UNWIND(	.cantunwind)
 
 	/* Resume into the restored TEE RAM */
 	ldr	r1, [r11, #PM_CTX_RESUME_PA]
+
+#ifdef CFG_PAGED_PSCI_SYSTEM_SUSPEND
+	/* Restore supervisor mode where pager is supported */
+	cps	#CPSR_MODE_SVC
+#endif
+
+	/* We're done */
 	bx	r1
 
 _failed:
@@ -694,12 +715,15 @@ level_done:
 	dsb	sy
 	isb
 
+#if !defined(CFG_PAGED_PSCI_SYSTEM_SUSPEND)
 	/* Resume sequence executes in Monitor mode */
 	cps	#CPSR_MODE_MON
+#endif
 
 	blx	plat_cpu_reset_early
 	b	sm_pm_cpu_resume
 END_FUNC stm32mp_sysram_resume
+DECLARE_KEEP_PAGER stm32mp_sysram_resume
 
 /*
  * stm32mp_cpu_reset_state - set CPU in a reset like state
@@ -708,8 +732,7 @@ END_FUNC stm32mp_sysram_resume
  * Preserve the execution mode in CPSR.
  */
 FUNC stm32mp_cpu_reset_state, :
-	push	{r12, lr}
-UNWIND(	.save	{r12, lr})
+	mov	r12, r0
 
 	cpsid	aif
 
@@ -725,5 +748,5 @@ UNWIND(	.save	{r12, lr})
 	dsb	sy
 	isb
 
-	pop	{r12, pc}
+	bx	r12
 END_FUNC stm32mp_cpu_reset_state
diff --git a/core/arch/arm/plat-stm32mp1/pm/psci.c b/core/arch/arm/plat-stm32mp1/pm/psci.c
index 49fb58d8c..5a74a77c5 100644
--- a/core/arch/arm/plat-stm32mp1/pm/psci.c
+++ b/core/arch/arm/plat-stm32mp1/pm/psci.c
@@ -25,6 +25,7 @@
 #include <kernel/pm.h>
 #include <kernel/spinlock.h>
 #include <mm/core_memprot.h>
+#include <pm/stm32mp1_psci.h>
 #include <platform_config.h>
 #include <sm/pm.h>
 #include <sm/psci.h>
@@ -330,9 +331,13 @@ static bool plat_can_suspend(void)
 	return rc;
 }
 
-/* Override default psci_system_suspend() with platform specific sequence */
-int psci_system_suspend(uintptr_t entry, uint32_t context_id __unused,
-			struct sm_nsec_ctx *nsec)
+/*
+ * Note: this function is weak just to make it possible to exclude it from
+ * the unpaged area.
+ */
+uint32_t __weak __psci_system_suspend(uintptr_t entry,
+				      uint32_t context_id __unused,
+				      struct sm_nsec_ctx *nsec)
 {
 	int ret = PSCI_RET_INVALID_PARAMETERS;
 	uint32_t soc_mode = 0;
@@ -361,7 +366,9 @@ int psci_system_suspend(uintptr_t entry, uint32_t context_id __unused,
 			return PSCI_RET_DENIED;
 #endif
 
-		sm_save_unbanked_regs(&nsec->ub_regs);
+		if (!IS_ENABLED(CFG_PAGED_PSCI_SYSTEM_SUSPEND))
+			sm_save_unbanked_regs(&nsec->ub_regs);
+
 		/*
 		 * sm_pm_cpu_suspend(arg, func) saves the CPU core context in
 		 * TEE RAM then calls func(arg) to run the platform lower power
@@ -377,7 +384,8 @@ int psci_system_suspend(uintptr_t entry, uint32_t context_id __unused,
 			stm32_exit_cstop();
 #endif
 			plat_resume((uint32_t)soc_mode);
-			sm_restore_unbanked_regs(&nsec->ub_regs);
+			if (!IS_ENABLED(CFG_PAGED_PSCI_SYSTEM_SUSPEND))
+				sm_restore_unbanked_regs(&nsec->ub_regs);
 		}
 	} else {
 		ret = plat_suspend((uint32_t)soc_mode);
@@ -392,8 +400,18 @@ int psci_system_suspend(uintptr_t entry, uint32_t context_id __unused,
 	return PSCI_RET_INTERNAL_FAILURE;
 }
 
-/* Override default psci_system_off() with platform specific sequence */
-void __noreturn psci_system_off(void)
+/* Override default psci_system_suspend() with platform specific sequence */
+int psci_system_suspend(uintptr_t entry, uint32_t context_id,
+			struct sm_nsec_ctx *nsec)
+{
+	return __psci_system_suspend(entry, context_id, nsec);
+}
+
+/*
+ * Note: this function is weak just to make it possible to exclude it from
+ * the unpaged area.
+ */
+void __weak __noreturn __psci_system_off(void)
 {
 	uint32_t soc_mode = stm32mp1_get_lp_soc_mode(PSCI_MODE_SYSTEM_OFF);
 
@@ -402,6 +420,11 @@ void __noreturn psci_system_off(void)
 	stm32_enter_cstop_shutdown(soc_mode);
 }
 
+void __noreturn psci_system_off(void)
+{
+	__psci_system_off();
+}
+
 /* Override default psci_system_reset() with platform specific sequence */
 void __noreturn psci_system_reset(void)
 {
diff --git a/core/arch/arm/plat-stm32mp1/pm/stm32mp1_psci.h b/core/arch/arm/plat-stm32mp1/pm/stm32mp1_psci.h
new file mode 100644
index 000000000..359587f8f
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp1/pm/stm32mp1_psci.h
@@ -0,0 +1,19 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+
+#ifndef __PM_STM32MP1_PSCI_H
+#define __PM_STM32MP1_PSCI_H
+
+#include <sm/sm.h>
+#include <stdint.h>
+
+/* STM32MP1 PSCI system suspend entry, possibly excluded from unpaged memory */
+uint32_t __psci_system_suspend(uintptr_t entry, uint32_t context_id,
+			       struct sm_nsec_ctx *nsec);
+
+/* STM32MP1 PSCI system off entry, possibly excluded from unpaged memory */
+void __psci_system_off(void);
+
+#endif /*__PM_STM32MP1_PSCI_H*/
-- 
2.47.2

